core-to-plc/src/Language/Plutus/CoreToPLC.hs

```diff
@@ -424,48 +546,62 @@ dataConCaseType resultType dc = withContextM (sdToTxt $ "Converting data constru
         -- t_1 -> ... -> t_m -> resultType
         pure $ mkIterTyFun args resultType
 
+dataConType :: Converting m => GHC.DataCon -> m PLCType
+dataConType dc =
+    let argTys = GHC.dataConRepArgTys dc
+        tc = GHC.dataConTyCon dc
+        tcTyVars = GHC.tyConTyVars tc
+    in
+        withContextM (sdToTxt $ "Converting data constructor type:" GHC.<+> GHC.ppr dc) $
+            flip (foldr (\tv acc -> mkTyForall tv acc)) tcTyVars $ do
+                args <- mapM convType argTys
+                resultType <- convType (GHC.dataConOrigResTy dc)
+                -- t_1 -> ... -> t_m -> resultType
+                pure $ mkIterTyFun args resultType
+
 -- This is the creation of the Scott-encoded constructor value.
-convConstructor :: Converting m => GHC.DataCon -> m PLCExpr
+convConstructor :: Converting m => GHC.DataCon -> m PLCTerm
 convConstructor dc =
     let
         tc = GHC.dataConTyCon dc
         tcTyVars = GHC.tyConTyVars tc
         tcName = GHC.getOccString $ GHC.getName tc
         dcName = GHC.getOccString $ GHC.getName dc
     in
-        -- no need for a body value check here, we know it's a lambda (see Note [Value restriction])
-        -- /\ tv_1 .. tv_n . body
-        flip (foldr (\tv acc -> mkTyAbs tv acc)) tcTyVars $ do
-            -- See Note [Scott encoding of datatypes]
-            resultType <- safeFreshTyName $ tcName ++ "_matchOut"
-            dcs <- getDataCons tc
-            index <- case elemIndex dc dcs of
-                Just i  -> pure i
-                Nothing -> throwPlain $ ConversionError "Data constructor not in the type constructor's list of constructors!"
-            caseTypes <- mapM (dataConCaseType (PLC.TyVar () resultType)) dcs
-            caseArgNames <- mapM (convNameFresh . GHC.getName) dcs
-            argTypes <- mapM convType $ GHC.dataConRepArgTys dc
-            argNames <- forM [0..(length argTypes -1)] (\i -> safeFreshName $ dcName ++ "_arg" ++ show i)
-
-            isSelfRecursive <- isSelfRecursiveTyCon tc
-            maybePf <- if isSelfRecursive
-                       then do
-                           tc' <- convTyCon tc
-                           case splitPatternFunctor tc' of
-                                Just (pf, n) -> pure $ Just (pf, n)
-                                Nothing -> throwPlain $ ConversionError "Could not compute pattern functor for recursive type"
-                       else pure Nothing
-
-            let scottBody = mkScottConstructorBody resultType (zip caseArgNames caseTypes) (zip argNames argTypes) index maybePf
-            pure scottBody
+        withContextM (sdToTxt $ "Converting data constructor:" GHC.<+> GHC.ppr dc) $
+            -- no need for a body value check here, we know it's a lambda (see Note [Value restriction])
+            -- /\ tv_1 .. tv_n . body
+            flip (foldr (\tv acc -> mkTyAbs tv acc)) tcTyVars $ do
```

Comment:
Add/use `mkIterTyAbs`? And perhaps give it (and others?) a different name, so they're not confused with stuff from `MkPlc`?